### PR TITLE
feat: add support for running the tests in ARM

### LINKED ISSUE #586: Add Linux arm as a target of e2e testing.
Allow the e2e testing to run on linux arm, we are currently targetting linux based host. 

- Figure out what distribution we need to support
- Figure out if anything need to change in the bootstrap.

Depends on  https://github.com/elastic/beats/issues/22893

### LINKED ISSUE #864: Support testing over linux-arm64.tar.gz, aarch64.rpm and arm64.deb artifacts (Docker handled separately)
Further towards https://github.com/elastic/observability-robots/issues/315
and 
https://github.com/elastic/beats/issues/22893
and
https://github.com/elastic/infra/issues/25762

we have not only the Docker image to validate, but 3 more arm64 linux artifacts to add to the mix.

### PR DESCRIPTION
**Pull request by mdelapenya**
## What does this PR do?
It enables the ARM platform for metricbeat and fleet test suites, grabbing ARM workers, and for that reason we are refactoring the way we define the architecture in the e2e code. Instead of hardcoding the `amd64` architecture, we are going to read it from the environment. Golang provides a `GOARCH` environment variable that is automatically set by Go's runtime. With that variable we are going to calculate the artifact name based on the underlying architecture:

- for CI workers running `amd64`, the code will generate an artifact for `amd64`.
- for CI workers running `arm64`, the code will generate an artifact for `arm64`.

Besides that, we had to rewrite the Jenkins pipeline to properly assign a worker in the right platform, to retrieve a non-immutable worker for `arm`.

Finally, we are adding the support to exclude scenarios per platform, leveraging the `.e2e-testing.yaml` file, being possiblee for a scenario to override its parent suite's platforms. To add awareness at this level, we are always removing the scenarios tagged with `@skip:PLATFORM` (amd64, arm64), including those tags in the feature file when needed, so that developers now what is going to be run directly in the source code, not in the CI descriptor.

## Why is it important?
Add support to ARM platform

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have run the Unit tests for the CLI, and they are passing locally
- [x] I have run the Unit tests for the E2E, and they are passing locally
- [x] I have run the End-2-End tests for the suite I'm working on, and they are passing locally
- [ ] I have noticed new Go dependencies (run `make notice` in the proper directory)

## Author's Checklist

- [x] Beats merged the compilation in `arm`.
- [ ] Official Ubi8 Docker images for `arm` are not in place. See https://github.com/elastic/beats/issues/24186. This will cause that any job for the E2E consuming an official release will fail the ubi8 tests for `arm`. The jobs triggered by Beats project (PRs and branches) won't be affected, as they consume the CI artifacts.
- [x] Beats-CI Ubi8 Docker images for `arm` are generated and pushed in our GCP bucket

## Related issues

- Closes #586
- Closes #864 
- Needs https://github.com/elastic/beats/issues/24186

## Screenshots

Thanks to the refactors we did in the past, it's easy to add parallel branches for new platforms, simply adding the Jenkins labels used by infra to grab the right CI workers.

Screenshot 2021-02-08 at 17 34 56

## Follow-ups
This PR is incomplete: let's see where it fails, as we need to update how the arch is calculated when creating the installers.

### CODE REVIEW

**Review by kuisathaverat — APPROVED** [review]

**Review by cachedout — APPROVED** [review]

File: .ci/Jenkinsfile
Comment (EricDavisX) [inline .ci/Jenkinsfile:320]: can you describe the intent of the branching?  I see it has to do with immutable workers, but don't have context - maybe that is what you could comment better on?

**Comment by EricDavisX** [comment]
@kvch fyi - Engg Productivity team is getting ready to hook our e2e-testing scenarios up for linux ARM testing.  awesome!

Thanks for putting the quick change in, Manu.  Question, I saw a note that we would need to wait until at least 7.12 BC2 (and after Noemi merges changes on Beats side, too) before we can hope to automate the scenario, let us all keep in touch.  :)

File: .ci/Jenkinsfile
Comment (mdelapenya) [reply .ci/Jenkinsfile:320]: Exactly, this is how Infra's workers are labelled: arm are not immutable yet

**Review by cachedout — APPROVED** [review]

File: .ci/Jenkinsfile
Comment (v1v) [inline .ci/Jenkinsfile:330]: * https://beats-ci.elastic.co/label/arm&&docker/
* https://beats-ci.elastic.co/label/immutable&&docker/

File: e2e/_suites/fleet/stand-alone.go
Comment (v1v) [inline e2e/_suites/fleet/stand-alone.go:79]: Docker images are pushed to the docker registry, and their tag name contains the architecture, for further details refers to https://github.com/elastic/beats/pull/24121

File: e2e/_suites/metricbeat/metricbeat_test.go
Comment (v1v) [inline e2e/_suites/metricbeat/metricbeat_test.go:413]: Docker images are pushed to the docker registry, and their tag name contains the architecture, for further details refers to elastic/beats#24121

**Review by v1v — APPROVED** [review]

File: .ci/Jenkinsfile
Comment (mdelapenya) [reply .ci/Jenkinsfile:330]: Will add this!

Comment (mdelapenya) [reply .ci/Jenkinsfile:330]: Mmm... does it mean that we cannot use `immutable` workers on ARM? If that's the case, the logic here is correct, right? we are only adding the immutable label for `!arm` workers.

File: e2e/_suites/fleet/stand-alone.go
Comment (mdelapenya) [inline e2e/_suites/fleet/stand-alone.go:103]:
```suggestion
	profileEnv["elasticAgentPlatform"] = "linux/" + arch
```

File: .ci/Jenkinsfile
Comment (mdelapenya) [inline .ci/Jenkinsfile:340]: @v1v do you think we can revert 0cd42e05badadaaa9e88c37ae4638033dcd3c8ff and apply that change to https://github.com/elastic/apm-pipeline-library/blob/6cfa9fac78ca050023ac0b5aa55de8d4090ff41c/vars/withGoEnvUnix.groovy#L48? Therefore, we will be setting the proper GOARCH value across any build.

**Review by cachedout — APPROVED** [review]

**Comment by mdelapenya** [comment]
https://github.com/elastic/beats/pull/25679 could affect the result of this build

**Comment by mdelapenya** [comment]
Ok, this seems to get good progress. The failures I can see seems related to certain flakiness that I need to investigate more in depth.

I.e: 

- https://beats-ci.elastic.co/blue/organizations/jenkins/e2e-tests%2Fe2e-testing-mbp/detail/PR-707/66/pipeline
- https://beats-ci.elastic.co/blue/organizations/jenkins/e2e-tests%2Fe2e-testing-mbp/detail/PR-707/65/pipeline
- https://beats-ci.elastic.co/blue/organizations/jenkins/e2e-tests%2Fe2e-testing-mbp/detail/PR-707/64/pipeline
- https://beats-ci.elastic.co/blue/organizations/jenkins/e2e-tests%2Fe2e-testing-mbp/detail/PR-707/63/pipeline

**Comment by mdelapenya** [comment]
Another topic to discuss is how we want to run the ARM tests: by PRs, on branches, on a nightly schedule...

**Comment by EricDavisX** [comment]
I spoke with the Endpoint security folk, and had been meaning to write.  We should disable the ARM tests that include Endpoint for now, it isn't a supported platform.  So... that will explain the consistent failure of the 2 tests:
Deploying an Endpoint makes policies to appear in the Security App
Changing an Agent policy is reflected in the Security App

That's great news, right?!  I'm sorry we didn't connect the errors to known support scope sooner.  

As 2nd, and new discussion point, we can acknowledge that tests for Elastic Endpoint in the context of Docker shouldn't really go very deep (as we have them coded).  The install / uninstall should work, theoretically, but it also, isn't a supported environment.  This is late in coming, we probably need to communicate better with the Endpoint team.  I'll take responsibility for that as well.  If we find significant flakiness with say Centos we can engage @ferullo to see if we can resolve it.  He notes he is happy to take a look at the endpoint logs and respond.  @mdelapenya do we have any current (otherwise stable) test runs that are showing flaky or failed tests for Centos wrt Endpoint ? 

Related, the Endpoint team is starting a significant effort to run the Endpoint on Containers and we'll need / want / get to update e2e-testing on Agent and Endpoint side over the next year (maybe within a few months we'll be able to discourse on specifics, but not just yet I don't think0.

**Comment by mdelapenya** [comment]
@EricDavisX we are going to add support to skip scenarios per platform, using this PR, thanks for your always helpful help :)

File: .ci/.e2e-tests.yaml
Comment (mdelapenya) [inline .ci/.e2e-tests.yaml:23]: It's possible to override platforms at the scenario level

File: .ci/Jenkinsfile
Comment (mdelapenya) [reply .ci/Jenkinsfile:332]: Will do this in a follow-up PR, thanks!

File: cli/config/compose/profiles/fleet/docker-compose.yml
Comment (mdelapenya) [inline cli/config/compose/profiles/fleet/docker-compose.yml:37]: If we get to an agreement to test PR versions of the package registry, then we can re-add this block again.

**Comment by mdelapenya** [comment]
@elastic/observablt-robots @EricDavisX I think the ARM support is feature-complete with this PR. Could you please 👀  so that we can merge?

File: .ci/Jenkinsfile
Comment (v1v) [inline .ci/Jenkinsfile:315]: The CI worker labels for the ARM64:
- 'aarch64'
- 'aws'
- 'ubuntu-1804-aarch64'
- 'arm' 

```suggestion
    workerLabels = "aarch64"
```

File: .ci/scripts/functional-test.sh
Comment (v1v) [inline .ci/scripts/functional-test.sh:32]: `GOARCH` is missing in the variable definitions, is that intended?

File: cli/config/compose/services/elastic-agent/centos/docker-compose.yml
Comment (v1v) [inline cli/config/compose/services/elastic-agent/centos/docker-compose.yml:7]: What's the reason for splitting the platform in elasticAgent and stack? IIUC, those variables are the same

File: .ci/scripts/functional-test.sh
Comment (mdelapenya) [reply .ci/scripts/functional-test.sh:32]: Mmm, this is set by the pipeline, within a `withEnv` block. Are you concerned about using this script outside the CI? If so, you are right, as the GOARCH is not set. Do you suggest passing it as a positional parameter instead? I think it would be a simple change.

File: cli/config/compose/services/elastic-agent/centos/docker-compose.yml
Comment (mdelapenya) [reply cli/config/compose/services/elastic-agent/centos/docker-compose.yml:7]: Yeah, it seems we can reuse it, as it's not possible to use different platforms. Will merge both variables into one in a new commit on top. Good catch!

Comment (mdelapenya) [reply cli/config/compose/services/elastic-agent/centos/docker-compose.yml:7]: In fact, I was thinking about removing the variable from compose files, and delegate the pull operation to Docker compose: as platform won't be set, docker compose will honour current host's platform (AMD/ARM), simplifying the configuration and code.

File: .ci/scripts/functional-test.sh
Comment (v1v) [reply .ci/scripts/functional-test.sh:32]: https://github.com/elastic/e2e-testing/pull/707/files#diff-e2a25c72a2cfde08a03b6735e6672600d998351ea4265e6b6146386aeb709e3bR17 uses that approach, that's the reason for my previous comment

Comment (mdelapenya) [reply .ci/scripts/functional-test.sh:32]: Ok, will proceed as in that commit. Thanks!!

File: .ci/Jenkinsfile
Comment (mdelapenya) [inline .ci/Jenkinsfile]: Temporarily reverting this, as I'm observing provisioning times for worker are slower, reaching build timeouts more frequently

**Comment by mdelapenya** [comment]
@adam-stokes waiting for your final review, and then this is ready to go!

**Review by adam-stokes — APPROVED** [review]

**Comment by adam-stokes** [comment]
@mdelapenya this should have minimal impact on my PR #1208 as I would only need to update the `internal/installer` code to make use of the architecture specific items. We can go ahead and merge this first and I'll update my PR on top of it.

**Comment by mdelapenya** [comment]
Let's merge this one once the regression with the integrations is resolved

**Comment by adam-stokes** [comment]
@mdelapenya this LGTM, just want to make sure you're ready for this to merge?

**Comment by mdelapenya** [comment]
> @mdelapenya this LGTM, just want to make sure you're ready for this to merge?

Yes! I did not mention it online, but I'd like to merge this today (cc/ @EricDavisX). We have been running this for days, twice, three, four times a day, and the results are consistent.

What I'd like to discuss with the @elastic/observablt-robots is about the schedule:
- should we run ARM tests with every PR (here or/and Beats/Kibana workflows)?

If we want to move the ARM tests to a less frequent schedule (because workers are more expensive), we could listen to _commands_ in the upstream repos to run a test suite in a platform, i.e. Fleet tests in ARM (plus the basic AMD tests).

**Comment by EricDavisX** [comment]
Thx Manu - exciting to have it merged.

**Pull request merged**
